README.md
# forsyde-sadf-examples

RISC processor and MPEG-4 decoder examples using ForSyDe-SADF.

## Note

This is a temporary repository for the case study for the paper

R. Bonna, D. S. Loubach, G. Ungureanu and I. Sander. _Modeling and Simulation of Dynamic Applications using Scenario-Aware Dataflow_

In the future these examples will be migrated and integrated to [`forsyde-shallow-examples`](https://github.com/forsyde/forsyde-shallow-examples).

## Installation

This project needs the [Haskell](https://www.haskell.org/) platform and the package manager [Stack](https://docs.haskellstack.org/en/stable/README/).

**Note:** install stack according to the [instructions on the web page](https://docs.haskellstack.org/en/stable/README/), i.e.

    curl -sSL https://get.haskellstack.org/ | sh

or

    wget -qO- https://get.haskellstack.org/ | sh

Update Stack, so that it reflects a current snapshot of the package repositories.

    stack upgrade
    stack update

Now you only need to install this project:

    cd path/to/forsydeSADF-examples
    stack install

and wait until everything is properly installed.

## Usage

The source files can be found in the [`src/`](src) directory. To load them into a Haskell interpreter session where you can execute all the declared functions individually, type in

    stack ghci

To run the provided test suites defined in in [`test/`](test/) with example input values, type in

    stack build :risc-example         # executes the RISC example in the paper
    stack build :mpeg-example         # executes the MPEG4 example with test input
    stack test                        # runs both previous examples




ChangeLog.md
# Changelog for forsyde-sadf-examples

## Unreleased changes


test/MpegExample.hs
module MpegExample where

import ForSyDe.Shallow
import Data.Matrix
import MPEG4

----------------------------------------------------------
--             MPEG4 SP Decoder parameters
----------------------------------------------------------

fs = (16,16)                      -- Frame size. Each element of fs should be a multiple of bs
bs = 8                            -- Macro blocks of bs x bs pixels
nb = div (uncurry (*) fs) (bs^2)  -- Number of macro blocks in a frame


----------------------------------------------------------
--                       Tests
----------------------------------------------------------

-- Before testing, remember to ignore the IDCT

frame1 = uncurry matrix fs $ \(i,j) -> 2*i + j
frame2 = identity (fst fs)
frame3 = uncurry fromList fs (repeat 1)

block1 = matrix bs bs $ \(i,j) -> (j-1)

iblocks = frame2mblocks (1,1) (bs,bs) frame2
p1block = FullB {pos = (3,3), block = block1, motionV = (0,-1)}
p2block1 = FullB {pos = (3,1), block = block1, motionV = (0,1)}
p2block2 = FullB {pos = (3,3), block = block1, motionV = (1,1)}

mvec = MotionVec { mvPos = (1, 1), mvVec = (-1, 0) }
mvecs = [mvec, mvec, mvec]

testFT = signal ["I", "P1", "P2", "P3"]
testinp = signal (iblocks ++ [p1block, p2block1, p2block2])

testout = pnMPEG4  bs nb fs testFT testinp

main = do
  putStrLn "\nMPEG test input blocks:\n"
  putStrLn "FT = "
  print testFT
  putStrLn "\nInput stream = "
  print testinp
  putStrLn "\nOutput frames:\n"
  print $ dropS 1 testout
  


test/RiscExample.hs
module RiscExample where

import ForSyDe.Shallow
import RISC

---------------------------------------------------------
-- Program Code
---------------------------------------------------------

progV :: Vector String
progV = vector [
  "movi 0 1",   -- reg 0 = 1
  "movi 1 1",   -- reg 1 = 1
  "movi 2 100", -- reg 2 = 100
  "add 0 1",    -- reg 0 = reg 0 + reg 1
  "sub 2 0",    -- reg 2 = reg 2 - reg 0
  "bez 2 1",    -- skip next instruction if reg 2 == 0
  "jmp -5",     -- jump back to instruction 3: "movi 2 100"
  "str 0 0",    -- mem(reg 0) = reg 0
  "ldr 10 0",   -- reg 10 = mem(reg 0)
  "outr 10",    -- output reg 10
  "outr 1",     -- output reg 1
  "outm 100",   -- output mem(100)
  "movi 10 7",  -- reg 10 = 7
  "st 10 42",   -- mem(42) = reg 10
  "outm 42",    -- output mem(42)
  "jmp -1"      -- end of program
  ]

progVcomments = [
  "\t# reg 0 = 1",
  "\t# reg 1 = 1",
  "\t# reg 2 = 100",
  "\t\t# reg 0 = reg 0 + reg 1",
  "\t\t# reg 2 = reg 2 - reg 0",
  "\t\t# skip next instruction if reg 2 == 0",
  "\t\t# jump back to instruction 3: \"movi 2 100\"",
  "\t\t# mem(reg 0) = reg 0",
  "\t# reg 10 = mem(reg 0)",
  "\t\t# output reg 10",
  "\t\t# output reg 1",
  "\t# output mem(100)",
  "\t# reg 10 = 7",
  "\t# mem(42) = reg 10",
  "\t\t# output mem(42)",
  "\t\t# end of program"
  ]


main = do
  putStrLn "\nRISC input program:\n"
  putStrLn $ unlines $ zipWith (\p c -> p ++ c) (fromVector progV) progVcomments
  putStrLn "\nOutput sequence:\n"
  print $ takeS 4 $ procNet progV
  


app/Main.hs
module Main where

import MPEG4

import ForSyDe.Shallow

import Text.Printf
import Control.Exception
import System.CPUTime
import Control.Monad
import Control.DeepSeq
import System.Environment
import Data.Matrix    -- use cabal install matrix to instal the matrix package

----------------------------------------------------------
--                       Tests
----------------------------------------------------------


lim :: Int
lim = 10^5

instance NFData a => NFData (Signal a) where rnf = rnf . fromSignal

time :: (NFData t) => Signal t -> IO (Double)
time y = do
    start <- getCPUTime
    replicateM_ lim $ do
        x <- evaluate $ y
        rnf x `seq` return ()
    end   <- getCPUTime
    let diff = (fromIntegral (end - start)) / (10^12)
    -- printf "Computation time: %0.9f sec\n" (diff :: Double)
    -- printf "Individual time: %0.9f sec\n" ( :: Double)
    return (diff / fromIntegral lim)

-- Before testing, remember to ignore the IDCT

ispow2 a | a == 2 && a > 0 = True
         | a > 0           = ispow2 (a `div` 2)
         | otherwise       = False

main = do
  a <- getArgs
  let [infsx, infsy, inbs, nSamp] = map read a :: [Int]
      -- read and check bs, fs and calculate nb
      bs | (ispow2 inbs) && (inbs <= 16) = inbs
      fs | infsx `mod` bs == 0 && infsy `mod` bs == 0 = (infsx,infsy)
      nb = div (uncurry (*) fs) (bs^2)
      -- create FT signal
      ftSig = signal $ cycle $ ["I"] ++ (map ((++) "P" . show) [1..(nb-1)])
      -- create input signal
      inSig = signal $ cycle $ genIBlocks ++ (concat . map genPBlocks) [1..(nb-1)]
      -- generator for I blocks
      genIBlocks = frame2mblocks (1,1) (bs,bs) idFrame
      idFrame = uncurry matrix fs $ \(i,j) -> if i == j then 1 else 0
      -- generator for P blocks
      genPBlocks x = take x $ cycle [pBlock1, pBlock2, pBlock3]
      blockDummy = matrix bs bs $ \(i,j) -> (j-1)
      pBlock1 = FullB {pos = (3,3), block = blockDummy, motionV = (0,-1)}
      pBlock2 = FullB {pos = (3,1), block = blockDummy, motionV = (0,1)}
      pBlock3 = FullB {pos = (3,3), block = blockDummy, motionV = (1,1)}
  -- start "real" computation
  let decoded = pnMPEG4 bs nb fs ftSig (takeS nSamp inSig)
  runtime <- time decoded
  -- print out the stats
  printf "%d\t%d\t%d\t%d\t%.9f\n" bs (fst fs * snd fs) nb nSamp runtime


src/RISC.hs
-----------------------------------------------------------------------------
-- |
-- Module  :  RISC_model
-- Copyright   :  (c) Ricardo Bonna
-- License     :  still needs license
--
-- Maintainer  :  ricardobonna@gmail.com
-- Stability   :  experimental
-- Portability :  portable
--
-- This is the second version of the RISC processor featuring a smaller
-- and more abstract concept. This makes easier to implement new operations.
--
-----------------------------------------------------------------------------

module RISC where

import ForSyDe.Shallow
import Data.Bits((.&.), (.|.), xor)


---------------------------------------------------------
-- Instruction Fetch (IF) Kernel
---------------------------------------------------------

type ScenarioIF = ((Int, Int, Int), (Int, Int, Int, Int), [Int] -> [Int]
                -> [Vector String] -> ([[Int]], [String], [Int], [Vector String]))

-- | 'ifScenario' is the list of possible scenarios for the IF kernel
ifScenario :: Int -> ScenarioIF
ifScenario n
  | n > 20 || n < 0   = error "ifScenario: Non existent scenario"
  | n >= 14 && n < 19 = ((1,1,1), (1,1,1,1), \[a] [pc] [m] -> ([arg m (pc+a)], [op m (pc+a)], [pc+a+1], [m])) -- branch
  | otherwise         = ((0,1,1), (1,1,1,1), \_ [pc] [m] -> ([arg m pc], [op m pc], [pc+1], [m]))             -- no branch

-- | 'slice' takes a string containing one assembly command and slice it into
-- one string with the opcode and a list of Ints containing the arguments
slice :: String -> (String, [Int])
slice a = (head $ words a, args)
  where rest = tail $ words a
        args
          | null rest = []
          | length rest == 1 = [read (head rest) :: Int]
          | length rest == 2 = [read (head rest) :: Int, read (last rest) :: Int]
          | otherwise = error "slice: Some instruction has more than 2 arguments"

arg :: (Eq a, Num a) => Vector String -> a -> [Int]
arg x y = snd $ slice (atV x y)

op :: (Eq a, Num a) => Vector String -> a -> String
op x y = fst $ slice (atV x y)

-- | 'ifKernel' is the Instruction Fetch (IF) kernel process
ifKernel :: Vector String -> Signal ScenarioIF -> Signal Int -> (Signal [Int], Signal String)
ifKernel progV ifCt sigBr = (sigArg, sigOp)
  where (sigArg, sigOp, sigPc, sigPm) = kernel34SADF ifCt sigBr sigPc' sigPm'
        sigPc' = delaySADF [0] sigPc
        sigPm' = delaySADF [progV] sigPm


---------------------------------------------------------
-- Execute (EXE) Kernel
---------------------------------------------------------

type ScenarioEXE = ((Int, Int, Int), (Int, Int, Int, Int), [[Int]] -> [Vector Int]
                 -> [Vector Int] -> ([Int], [Int], [Vector Int], [Vector Int]))

-- | 'exeScenario' is the list of possible scenarios for the EXE kernel
exeScenario :: Int -> ScenarioEXE
exeScenario 0  = ((1,0,0), (0,0,0,0), \_ _ _ -> ([], [], [], []))                                                     -- nop
exeScenario 1  = ((1,1,1), (0,0,1,1), \[[rd, mn]] [r] [m] -> ([], [], [replaceV r rd (atV m mn)], [m]))               -- ld
exeScenario 2  = ((1,1,1), (0,0,1,1), \[[rd, rm]] [r] [m] -> ([], [], [replaceV r rd (atV m (atV r rm))], [m]))       -- ldr
exeScenario 3  = ((1,1,1), (0,0,1,1), \[[rs, mn]] [r] [m] -> ([], [], [r], [replaceV m mn (atV r rs)]))               -- st
exeScenario 4  = ((1,1,1), (0,0,1,1), \[[rs, rm]] [r] [m] -> ([], [], [r], [replaceV m (atV r rm) (atV r rs)]))       -- str
exeScenario 5  = ((1,1,0), (0,0,1,0), \[[rd, rs]] [r] _ -> ([], [], [replaceV r rd (atV r rs)], []))                  -- mov
exeScenario 6  = ((1,1,0), (0,0,1,0), \[[rd, i]] [r] _ -> ([], [], [replaceV r rd i], []))                            -- movi
exeScenario 7  = ((1,1,0), (0,0,1,0), \[[rd, rs]] [r] _ -> ([], [], [replaceV r rd ((atV r rd) + (atV r rs))], []))   -- add
exeScenario 8  = ((1,1,0), (0,0,1,0), \[[rd, rs]] [r] _ -> ([], [], [replaceV r rd ((atV r rd) - (atV r rs))], []))   -- sub
exeScenario 9  = ((1,1,0), (0,0,1,0), \[[rd, rs]] [r] _ -> ([], [], [replaceV r rd ((atV r rd) * (atV r rs))], []))   -- mul
exeScenario 10 = ((1,1,0), (0,0,1,0), \[[rd, rs]] [r] _ -> ([], [], [replaceV r rd (div (atV r rd) (atV r rs))], [])) -- div
exeScenario 11 = ((1,1,0), (0,0,1,0), \[[rd, rs]] [r] _ -> ([], [], [replaceV r rd ((atV r rd) .&. (atV r rs))], [])) -- and
exeScenario 12 = ((1,1,0), (0,0,1,0), \[[rd, rs]] [r] _ -> ([], [], [replaceV r rd ((atV r rd) .|. (atV r rs))], [])) -- or
exeScenario 13 = ((1,1,0), (0,0,1,0), \[[rd, rs]] [r] _ -> ([], [], [replaceV r rd (xor (atV r rd) (atV r rs))], [])) -- xor
exeScenario 14 = ((1,1,0), (1,0,1,0), \[[rs, v]] [r] _ -> ([if atV r rs == 0 then v else 0], [], [r], []))            -- bez
exeScenario 15 = ((1,1,0), (1,0,1,0), \[[rs, v]] [r] _ -> ([if atV r rs /= 0 then v else 0], [], [r], []))            -- bnz
exeScenario 16 = ((1,1,0), (1,0,1,0), \[[rs, v]] [r] _ -> ([if atV r rs > 0 then v else 0], [], [r], []))             -- bgz
exeScenario 17 = ((1,1,0), (1,0,1,0), \[[rs, v]] [r] _ -> ([if atV r rs < 0 then v else 0], [], [r], []))             -- blz
exeScenario 18 = ((1,0,0), (1,0,0,0), \[[v]] _ _ -> ([v], [], [], []))                                                -- jmp
exeScenario 19 = ((1,1,0), (0,1,1,0), \[[rs]] [r] _ -> ([], [atV r rs], [r], []))                                     -- outr
exeScenario 20 = ((1,0,1), (0,1,0,1), \[[mn]] _ [m] -> ([], [atV m mn], [], [m]))                                     -- outm
exeScenario _  = error "exeScenario: Non existent scenario"

-- | 'exeKernel' is the Execution (EXE) kernel process
exeKernel :: Signal ScenarioEXE -> Signal [Int] -> (Signal Int, Signal Int)
exeKernel exeCt sigArg = (sigBr, sigDmp)
  where (sigBr, sigDmp, sigReg, sigDm) = kernel34SADF exeCt sigArg sigReg' sigDm'
        sigReg' = delaySADF [regV] sigReg
        sigDm' = delaySADF [memV] sigDm

regV = vector $ replicate 32 0
memV = vector $ replicate 1024 0

---------------------------------------------------------
-- Decode Detector
---------------------------------------------------------

type ScenarioDEC = ((Int,Int), ([ScenarioIF], [ScenarioEXE]))

type StateDEC = Int

-- | 'detectorScenario' is th output function of the detector. It converts a
-- state into an equivalent scenario
decScenario :: StateDEC -> ScenarioDEC
decScenario n = ((1,1), ([ifScenario n], [exeScenario n]))

-- | 'decSwitchState' is the state transition function of the detector
decSwitchState :: StateDEC -> [String] -> StateDEC
decSwitchState _ ["nop"]  = 0
decSwitchState _ ["ld"]   = 1
decSwitchState _ ["ldr"]  = 2
decSwitchState _ ["st"]   = 3
decSwitchState _ ["str"]  = 4
decSwitchState _ ["mov"]  = 5
decSwitchState _ ["movi"] = 6
decSwitchState _ ["add"]  = 7
decSwitchState _ ["sub"]  = 8
decSwitchState _ ["mul"]  = 9
decSwitchState _ ["div"]  = 10
decSwitchState _ ["and"]  = 11
decSwitchState _ ["or"]   = 12
decSwitchState _ ["xor"]  = 13
decSwitchState _ ["bez"]  = 14
decSwitchState _ ["bnz"]  = 15
decSwitchState _ ["bgz"]  = 16
decSwitchState _ ["blz"]  = 17
decSwitchState _ ["jmp"]  = 18
decSwitchState _ ["outr"] = 19
decSwitchState _ ["outm"] = 20
decSwitchState _ _ = error "decSwitchState: Input not recognized"

-- | 'decodeDetector' is the detector of ProSyDe
decDetector :: Signal String -> (Signal ScenarioIF, Signal ScenarioEXE)
decDetector = detector12SADF 1 decSwitchState decScenario 0


---------------------------------------------------------
-- ProSyDe compact
---------------------------------------------------------

-- | 'prosyde' is the processor built with the ForSyDe SADF library. It outputs
-- whatever it is in register @rs@ during the execution of instruction @outr rs@
procNet :: Vector String -> Signal Int
procNet progV = sigDmp
  where (sigBr, sigDmp) = exeKernel exeCt sigArg
        (sigArg, sigOp) = ifKernel progV ifCt' sigBr
        (ifCt, exeCt) = decDetector sigOp
        ifCt' = delaySADF [ifScenario 0] ifCt




examples/RISC_model.hs
-----------------------------------------------------------------------------
-- |
-- Module  :  RISC_model
-- Copyright   :  (c) Ricardo Bonna
-- License     :  still needs license
--
-- Maintainer  :  ricardobonna@gmail.com
-- Stability   :  experimental
-- Portability :  portable
--
-- This is the second version of the RISC processor featuring a smaller
-- and more abstract concept. This makes easier to implement new operations.
--
-----------------------------------------------------------------------------

module RISC_model (
  procNet
  ) where

import ForSyDe.Shallow
import Data.Bits((.&.), (.|.), xor)


---------------------------------------------------------
-- Instruction Fetch (IF) Kernel
---------------------------------------------------------

type ScenarioIF = ((Int, Int, Int), (Int, Int, Int, Int), [Int] -> [Int]
                -> [Vector String] -> ([[Int]], [String], [Int], [Vector String]))

-- | 'ifScenario' is the list of possible scenarios for the IF kernel
ifScenario :: Int -> ScenarioIF
ifScenario n
  | n > 20 || n < 0   = error "ifScenario: Non existent scenario"
  | n >= 14 && n < 19 = ((1,1,1), (1,1,1,1), \[a] [pc] [m] -> ([arg m (pc+a)], [op m (pc+a)], [pc+a+1], [m])) -- branch
  | otherwise         = ((0,1,1), (1,1,1,1), \_ [pc] [m] -> ([arg m pc], [op m pc], [pc+1], [m]))             -- no branch

-- | 'slice' takes a string containing one assembly command and slice it into
-- one string with the opcode and a list of Ints containing the arguments
slice :: String -> (String, [Int])
slice a = (head $ words a, args)
  where rest = tail $ words a
        args
          | null rest = []
          | length rest == 1 = [read (head rest) :: Int]
          | length rest == 2 = [read (head rest) :: Int, read (last rest) :: Int]
          | otherwise = error "slice: Some instruction has more than 2 arguments"

arg :: (Eq a, Num a) => Vector String -> a -> [Int]
arg x y = snd $ slice (atV x y)

op :: (Eq a, Num a) => Vector String -> a -> String
op x y = fst $ slice (atV x y)

-- | 'ifKernel' is the Instruction Fetch (IF) kernel process
ifKernel :: Vector String -> Signal ScenarioIF -> Signal Int -> (Signal [Int], Signal String)
ifKernel progV ifCt sigBr = (sigArg, sigOp)
  where (sigArg, sigOp, sigPc, sigPm) = kernel34SADF ifCt sigBr sigPc' sigPm'
        sigPc' = delaySADF [0] sigPc
        sigPm' = delaySADF [progV] sigPm


---------------------------------------------------------
-- Execute (EXE) Kernel
---------------------------------------------------------

type ScenarioEXE = ((Int, Int, Int), (Int, Int, Int, Int), [[Int]] -> [Vector Int]
                 -> [Vector Int] -> ([Int], [Int], [Vector Int], [Vector Int]))

-- | 'exeScenario' is the list of possible scenarios for the EXE kernel
exeScenario :: Int -> ScenarioEXE
exeScenario 0  = ((1,0,0), (0,0,0,0), \_ _ _ -> ([], [], [], []))                                                     -- nop
exeScenario 1  = ((1,1,1), (0,0,1,1), \[[rd, mn]] [r] [m] -> ([], [], [replaceV r rd (atV m mn)], [m]))               -- ld
exeScenario 2  = ((1,1,1), (0,0,1,1), \[[rd, rm]] [r] [m] -> ([], [], [replaceV r rd (atV m (atV r rm))], [m]))       -- ldr
exeScenario 3  = ((1,1,1), (0,0,1,1), \[[rs, mn]] [r] [m] -> ([], [], [r], [replaceV m mn (atV r rs)]))               -- st
exeScenario 4  = ((1,1,1), (0,0,1,1), \[[rs, rm]] [r] [m] -> ([], [], [r], [replaceV m (atV r rm) (atV r rs)]))       -- str
exeScenario 5  = ((1,1,0), (0,0,1,0), \[[rd, rs]] [r] _ -> ([], [], [replaceV r rd (atV r rs)], []))                  -- mov
exeScenario 6  = ((1,1,0), (0,0,1,0), \[[rd, i]] [r] _ -> ([], [], [replaceV r rd i], []))                            -- movi
exeScenario 7  = ((1,1,0), (0,0,1,0), \[[rd, rs]] [r] _ -> ([], [], [replaceV r rd ((atV r rd) + (atV r rs))], []))   -- add
exeScenario 8  = ((1,1,0), (0,0,1,0), \[[rd, rs]] [r] _ -> ([], [], [replaceV r rd ((atV r rd) - (atV r rs))], []))   -- sub
exeScenario 9  = ((1,1,0), (0,0,1,0), \[[rd, rs]] [r] _ -> ([], [], [replaceV r rd ((atV r rd) * (atV r rs))], []))   -- mul
exeScenario 10 = ((1,1,0), (0,0,1,0), \[[rd, rs]] [r] _ -> ([], [], [replaceV r rd (div (atV r rd) (atV r rs))], [])) -- div
exeScenario 11 = ((1,1,0), (0,0,1,0), \[[rd, rs]] [r] _ -> ([], [], [replaceV r rd ((atV r rd) .&. (atV r rs))], [])) -- and
exeScenario 12 = ((1,1,0), (0,0,1,0), \[[rd, rs]] [r] _ -> ([], [], [replaceV r rd ((atV r rd) .|. (atV r rs))], [])) -- or
exeScenario 13 = ((1,1,0), (0,0,1,0), \[[rd, rs]] [r] _ -> ([], [], [replaceV r rd (xor (atV r rd) (atV r rs))], [])) -- xor
exeScenario 14 = ((1,1,0), (1,0,1,0), \[[rs, v]] [r] _ -> ([if atV r rs == 0 then v else 0], [], [r], []))            -- bez
exeScenario 15 = ((1,1,0), (1,0,1,0), \[[rs, v]] [r] _ -> ([if atV r rs /= 0 then v else 0], [], [r], []))            -- bnz
exeScenario 16 = ((1,1,0), (1,0,1,0), \[[rs, v]] [r] _ -> ([if atV r rs > 0 then v else 0], [], [r], []))             -- bgz
exeScenario 17 = ((1,1,0), (1,0,1,0), \[[rs, v]] [r] _ -> ([if atV r rs < 0 then v else 0], [], [r], []))             -- blz
exeScenario 18 = ((1,0,0), (1,0,0,0), \[[v]] _ _ -> ([v], [], [], []))                                                -- jmp
exeScenario 19 = ((1,1,0), (0,1,1,0), \[[rs]] [r] _ -> ([], [atV r rs], [r], []))                                     -- outr
exeScenario 20 = ((1,0,1), (0,1,0,1), \[[mn]] _ [m] -> ([], [atV m mn], [], [m]))                                     -- outm
exeScenario _  = error "exeScenario: Non existent scenario"

-- | 'exeKernel' is the Execution (EXE) kernel process
exeKernel :: Signal ScenarioEXE -> Signal [Int] -> (Signal Int, Signal Int)
exeKernel exeCt sigArg = (sigBr, sigDmp)
  where (sigBr, sigDmp, sigReg, sigDm) = kernel34SADF exeCt sigArg sigReg' sigDm'
        sigReg' = delaySADF [regV] sigReg
        sigDm' = delaySADF [memV] sigDm

regV = vector $ replicate 32 0
memV = vector $ replicate 1024 0

---------------------------------------------------------
-- Decode Detector
---------------------------------------------------------

type ScenarioDEC = ((Int,Int), ([ScenarioIF], [ScenarioEXE]))

type StateDEC = Int

-- | 'detectorScenario' is th output function of the detector. It converts a
-- state into an equivalent scenario
decScenario :: StateDEC -> ScenarioDEC
decScenario n = ((1,1), ([ifScenario n], [exeScenario n]))

-- | 'decSwitchState' is the state transition function of the detector
decSwitchState :: StateDEC -> [String] -> StateDEC
decSwitchState _ ["nop"]  = 0
decSwitchState _ ["ld"]   = 1
decSwitchState _ ["ldr"]  = 2
decSwitchState _ ["st"]   = 3
decSwitchState _ ["str"]  = 4
decSwitchState _ ["mov"]  = 5
decSwitchState _ ["movi"] = 6
decSwitchState _ ["add"]  = 7
decSwitchState _ ["sub"]  = 8
decSwitchState _ ["mul"]  = 9
decSwitchState _ ["div"]  = 10
decSwitchState _ ["and"]  = 11
decSwitchState _ ["or"]   = 12
decSwitchState _ ["xor"]  = 13
decSwitchState _ ["bez"]  = 14
decSwitchState _ ["bnz"]  = 15
decSwitchState _ ["bgz"]  = 16
decSwitchState _ ["blz"]  = 17
decSwitchState _ ["jmp"]  = 18
decSwitchState _ ["outr"] = 19
decSwitchState _ ["outm"] = 20
decSwitchState _ _ = error "decSwitchState: Input not recognized"

-- | 'decodeDetector' is the detector of ProSyDe
decDetector :: Signal String -> (Signal ScenarioIF, Signal ScenarioEXE)
decDetector = detector12SADF 1 decSwitchState decScenario 0


---------------------------------------------------------
-- ProSyDe compact
---------------------------------------------------------

-- | 'prosyde' is the processor built with the ForSyDe SADF library. It outputs
-- whatever it is in register @rs@ during the execution of instruction @outr rs@
procNet :: Vector String -> Signal Int
procNet progV = sigDmp
  where (sigBr, sigDmp) = exeKernel exeCt sigArg
        (sigArg, sigOp) = ifKernel progV ifCt' sigBr
        (ifCt, exeCt) = decDetector sigOp
        ifCt' = delaySADF [ifScenario 0] ifCt


---------------------------------------------------------
-- Program Code
---------------------------------------------------------

progV :: Vector String
progV = vector [
  "movi 0 1",   -- reg 0 = 1
  "movi 1 1",   -- reg 1 = 1
  "movi 2 100", -- reg 2 = 100
  "add 0 1",    -- reg 0 = reg 0 + reg 1
  "sub 2 0",    -- reg 2 = reg 2 - reg 0
  "bez 2 1",    -- skip next instruction if reg 2 == 0
  "jmp -5",     -- jump back to instruction 3: "movi 2 100"
  "str 0 0",    -- mem(reg 0) = reg 0
  "ldr 10 0",   -- reg 10 = mem(reg 0)
  "outr 10",    -- output reg 10
  "outr 1",     -- output reg 1
  "outm 100",   -- output mem(100)
  "movi 10 7",  -- reg 10 = 7
  "st 10 42",   -- mem(42) = reg 10
  "outm 42",    -- output mem(42)
  "jmp -1"      -- end of program
  ]
